### DISCUSSION TITLE
Do API routes in Next.js support native node modules?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#34040] [Help]
Hello,

I'm trying to use native node modules with an API route. I tried editing my webpack configuration like so using the [node-loader](https://webpack.js.org/loaders/node-loader/#root):

```js
const withPlugins = require('next-compose-plugins')
const withMDX = require('@next/mdx')

module.exports = withPlugins([
  withMDX({ providerImportSource: '@mdx-js/react' })
], {
  webpack: (config,{ isServer }) => {
    if (isServer) {
      config.module.rules.push({
        test: /\.node$/,
        use: [
          {
            loader: 'node-loader',
            options: {
              name: '[path][name].[ext]'
            }
          }
        ]
      })
    }

    return config
  }
})
```

However, I then got a `(no such file)` error like the following:

```
Error: dlopen(/site/website/.next/server/pages/api/_/_/_/_/node_modules/neon-node-package/index.node, 0x0001): tried: '/site/website/.next/server/pages/api/_/_/_/_/node_modules/neon-node-package/index.node' (no such file), '/usr/local/lib/index.node' (no such file), '/usr/lib/index.node' (no such file)
```

Without the node loader, I got an error saying that my application could not handle binary files. Has anyone tried doing this before? I'd love to use native node modules built with [Neon Bindings](https://neon-bindings.com) in my Next.js App.

**Comment by ghost** [comment]
If I try using the `defaultLoaders.babel` I get an error w/ the app trying to use babel to run a binary file. Is there a way to simply copy the files from `node_modules` to a `.next` folder for the page itself and then pipe that to the `node-loader`?

**Comment by jkhoffman** [comment]
Same issue here.

**Comment by amarjanica** [comment]
Same issue. I tried using the node-loader, but not working as expected.
One alternative would be creating a serverless function or creating a separate node server, but that's out of this scope and an overkill.

The lookup path is wrong. Server tries to require the module relative to the api route directory.
What I mean for example:
*pages/api/hello-world.ts*
```
import slugify from 'node-loader!node-el-slugify';

export default async function handler(req, res) {
  res.status(200).json({ message: slugify.slugify('Hello World!')})
}
```
And calling the route breaks with:
```
Error: node-loader: ~/myproject/.next/server/pages/api/hello-world/fd9b94534ce9afecec2d9988e8166596.node: cannot open shared object file: No such file or directory
```
If I manually copy the module, then it works.
I think this should be easy to fix somewhere in the webpack config, but I'm lost.

**Reply by amarjanica** [reply]
@shicholas you've probably found a solution by now, but writing this for anyone who stumbles upon this question.
API routes in next.js support native node modules, just configure webpack to use https://github.com/vercel/webpack-asset-relocator-loader.

**Reply by levkk** [reply]
It seems to work for `npm run dev`, but when building a production build, it doesn't relocate the `.node` modules into `.next/server` and I'm getting a ` Cannot find module` error nonetheless.

**Reply by StringKe** [reply]
The server-side compilation of nextjs merges multiple files, which results in inaccurate paths after being processed by the node-loader.

```
config.module.rules.push({
                test: /\.node$/,
                use: [
                    {
                        loader: 'node-loader',
                        options: {
                            name: '[name].[ext]',
                        },
                    },
                ],
});
```

**Reply by StringKe** [reply]
@amarjanica I googled your program and it worked for me, thank you!

Anyone having the same problem can use https://github.com/eisberg-labs/nextjs-node-loader

**Reply by emanueleecho** [reply]
hey mate i tried to use it but i'm getting this error
- error Conflict: Multiple assets emit different content to the same filename koffi.node
can anyone help?

**Reply by amarjanica** [reply]
Hi @emanueleecho I got your bug report, let's continue the discussion here https://github.com/eisberg-labs/nextjs-node-loader/issues/18

**Comment by amarjanica** [comment]
Hey, I don't like leaving this unanswered. There's no need to use nextjs-node-loader. 
Solution in NextJS 14 is to add a: 
```
experimental: {
    serverComponentsExternalPackages: ['yournativemodule'],
}
```
which will be in NextJS 15 just:
```
serverExternalPackages: ['yournativemodule']
```
I had an issue with a dependency that was built with neon binding, so an additional configuration was:
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    serverComponentsExternalPackages: ['mynativemodule'],
  },
  /** @type {import('webpack').Configuration} */
  webpack: (config, context) => {
    if (context.isServer) {
      config.externals = [
        ...config.externals,
        {'mynativemodule': 'commonjs mynativemodule'},
      ]
    }
    return config;
  },
};
```
This configuration tells Webpack to treat `mynativemodule` as an external module and not bundle it. `externals` wasn't needed for other native modules, just the ones built with neon binding.
